packages/itwin/grouping-mapping-widget/src/test/GroupingMappingCustomUI.test.tsx

```diff
@@ -0,0 +1,195 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import React from "react";
+import "@testing-library/jest-dom";
+import { render, screen, TestUtils, waitForElementToBeRemoved, within } from "./test-utils";
+import { faker } from "@faker-js/faker";
+import { Groupings } from "../grouping-mapping-widget";
+import type { GroupCollection, Mapping } from "@itwin/insights-client";
+import * as moq from "typemoq";
+import type { IModelConnection, SelectionSet, SelectionSetEvent} from "@itwin/core-frontend";
+import { NoRenderApp } from "@itwin/core-frontend";
+import type { SelectionManager, SelectionScopesManager } from "@itwin/presentation-frontend";
+import { Presentation, SelectionChangeEvent } from "@itwin/presentation-frontend";
+import type { BeEvent } from "@itwin/core-bentley";
+import { setupServer } from "msw/node";
+import { rest } from "msw";
+
+const mockITwinId = faker.datatype.uuid();
+const mockIModelId = faker.datatype.uuid();
+const mockMappingId = faker.datatype.uuid();
+const mockMapping: Mapping = {
+  id: mockMappingId,
+  mappingName: "mOcKmApPiNg1",
+  description: "mOcKmApPiNgDeScRiPtIoN1",
+  createdBy: faker.random.alpha(),
+  createdOn: faker.date.past().toDateString(),
+  modifiedBy: faker.random.alpha(),
+  modifiedOn: faker.date.past().toDateString(),
+  extractionEnabled:false,
+  _links: {
+    imodel: {
+      href: "",
+    },
+  },
+};
+
+const groupsFactory = (): GroupCollection => ({
+  groups: Array.from(
+    { length: faker.datatype.number({ min: 3, max: 5 }) },
+    (_, index) => ({
+      id: `${faker.datatype.uuid()}`,
+      groupName: `mOcKgRoUp${index}`,
+      description: `mOcKgRoUpDeScRiPtIoN${index}`,
+      query: `mOcKgRoUpQuErY${index}`,
+      deleted: false,
+      _links: {
+        imodel: {
+          href: "",
+        },
+        mapping: {
+          href: "",
+        },
+      },
+    })
+  ),
+  _links: {
+    next: undefined,
+    self: {
+      href: "",
+    },
+  },
+});
+
+const connectionMock = moq.Mock.ofType<IModelConnection>();
+const selectionManagerMock = moq.Mock.ofType<SelectionManager>();
+const selectionScopesManagerMock = moq.Mock.ofType<SelectionScopesManager>();
+
+jest.mock("@itwin/appui-react", () => ({
+  ...jest.requireActual("@itwin/appui-react"),
+  useActiveIModelConnection: () => connectionMock.object,
+}));
+
+const server = setupServer();
+
+beforeAll(async () => {
```

Comment:
This `beforeAll` will probably run before ALL tests in this test suite. I'd say it should be scoped to the specific component being tested. Use `describe` to reduce the scope. Also, mocks should be set up in that scope as well.

Reply:
Wouldn't it be better to be kept in beforeAll so that it only runs once for entire suite? Each test will need to render the components within widget and will have interfaction with viewer

Reply:
Having global setups can make it harder to debug other tests. I think you can run into some weird race conditions if you keep it this way. If you really want to have a global setup for all tests - move it into a separate file. The require it for the test script in package.json (e.g. https://github.com/iTwin/insights-client/blob/main/package.json, see `--require` for `test:integration`).

Reply:
I looked into jest configuration. The beforeAll is scoped to each test file so there won't be a global race condition. There is a jest configuration 'globalSetup' which will be called before all test suites but it is run before test-environment is setup. I tried it but a lot of setups here involves jsdom and it doesn't work. I submitted an issue in Jest repository and wait for anybody to give me a solution. Right now, I don't think it worths the time to dig deeper into this problem. We will have more bandwidth in Q4 for GM testing